### PR TITLE
feat: adapter whitelist

### PR DESCRIPTION
**Pull request by MathisGD**
TODO

### CODE REVIEW

**Comment by qdrn** [comment]
One "weird" thing is that you can add adapters and then a whitelister and end up in a state where none of the adapters are whitelisted...

**Comment by MattLsb** [comment]
>  If the old adapter is removed, existing vaults can't use the new adapter without removing the old adapter (which is extremely heavy)
I’m not sure I understand. There’s no requirement to remove the old adapter. The issue only comes up if a new registry is added without including the old adapter, not when the existing registry is updated.

If there’s no critical issue with the old adapter, it’s fine for the registry to list both the old and new adapters.

That said, if an old adapter does have a serious problem, there’s currently no way to enforce aligned vaults to remove it via policy (and none of the proposed options solve this).

**Comment by MattLsb** [comment]
I agree. Note that this makes checking whether a vault is ok a bit more complex.

For example in the app:

check if the registry is Morpho and addRegister is abdicated
vs
check the vault's adapters and compare them against the Morpho registry (past and present).

Still seems fine overall though.

**Comment by MathisGD** [comment]
convinced that we should leave the behavior as is. The main argument is that constraining future adapters is useful for businesses (like distributors or morpho itself) but not users, so it's fine to have a bit more complexity in the use (to check the "alignment").

**Comment by peyha** [comment]
The Readme could be updated in the adapter/curator part (I feel like we stopped updating it recently?)

File: src/interfaces/IAdapterRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (c) 2025 Morpho Association
+pragma solidity >=0.5.0;
+
+interface IAdapterRegistry {
+    function isValidAdapter(address account) external view returns (bool);
```
Comment (MathisGD) [inline src/interfaces/IAdapterRegistry.sol:6]: Wdyt?
```suggestion
    function isInRegistry(address account) external view returns (bool);
```

Comment (adhusson) [reply src/interfaces/IAdapterRegistry.sol:6]: Will update

**Review by peyha — APPROVED** [review]
what was the reason for checking all current adapter in `setAdapterRegistry` in the end ?

**Comment by adhusson** [comment]
> what was the reason for checking all current adapter in setAdapterRegistry in the end ?

So you have the property that if a vault uses an add-only adapter registry (a registry that never un-whitelists an adapter), that vault can only ever have adapters approved by the registry.

It means fewer things to check from the outside.

File: test/SettersTest.sol [hunk 2]
```diff
@@ -239,6 +239,26 @@ contract SettersTest is BaseTest {
         assertEq(vault.adapters(0), newAdapter);
     }
 
+    function testAddAdapterRegistry(bool isInRegistry) public {
```
Comment (peyha) [inline test/SettersTest.sol:242]: a bit weird that this bool has the same name as the function in `IAdapterRegistry` but no big deal

**Review by peyha — APPROVED** [review]
~~we could add a test when an adapter is set in the registry and then unset to check that vaults revert but it's an edge case so not mandatory imo~~
